File: android/app/src/main/java/com/openledger/android/ui/auth/LoginViewModel.kt
package com.openledger.android.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openledger.android.data.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class LoginState(
    val email: String = "",
    val password: String = "",
    val isLoading: Boolean = false,
    val error: String? = null,
)

@HiltViewModel
class LoginViewModel @Inject constructor(
    private val authRepo: AuthRepository,
) : ViewModel() {

    private val _state = MutableStateFlow(LoginState())
    val state: StateFlow<LoginState> = _state.asStateFlow()

    fun onEmailChanged(email: String) {
        _state.value = _state.value.copy(email = email, error = null)
    }

    fun onPasswordChanged(password: String) {
        _state.value = _state.value.copy(password = password, error = null)
    }

    fun login(onSuccess: () -> Unit) {
        _state.value = _state.value.copy(isLoading = true, error = null)
        viewModelScope.launch {
            authRepo.login(_state.value.email, _state.value.password)
                .onSuccess { onSuccess() }
                .onFailure { e ->
                    _state.value = _state.value.copy(
                        isLoading = false,
                        error = "Login failed: ${e.message}",
                    )
                }
        }
    }
}

File: ios/OpenLedger/OpenLedger/ViewModels/LoginViewModel.swift
import Foundation

@Observable
final class LoginViewModel {

    var email: String = ""
    var password: String = ""
    var isLoading: Bool = false
    var error: String?

    func login(authRepo: AuthRepository, onSuccess: @escaping () -> Void) {
        guard !email.trimmingCharacters(in: .whitespaces).isEmpty,
              !password.isEmpty else {
            error = "Please enter email and password."
            return
        }

        isLoading = true
        error = nil

        Task { @MainActor in
            do {
                try await authRepo.login(email: email, password: password)
                isLoading = false
                onSuccess()
            } catch {
                isLoading = false
                self.error = "Login failed: \(error.localizedDescription)"
            }
        }
    }
}
